addons/binding/org.openhab.binding.kodi/src/main/java/org/openhab/binding/kodi/internal/protocol/KodiConnection.java

```diff
@@ -332,6 +328,41 @@ private String convertToText(JsonElement element) {
         // }
     }
 
+    private String convertToImageUrl(JsonElement element) {
+        String text = convertToText(element);
+        if (!text.isEmpty()) {
+            try {
+                // we have to strip ending "/" here because Kodi returns a not valid path and filename
+                // "fanart":"image://http%3a%2f%2fthetvdb.com%2fbanners%2ffanart%2foriginal%2f263365-31.jpg/"
+                // "thumbnail":"image://http%3a%2f%2fthetvdb.com%2fbanners%2fepisodes%2f263365%2f5640869.jpg/"
+                String encodedURL = URLEncoder.encode(StringUtils.stripEnd(text, "/"), "UTF-8");
+                return imageUri.resolve(encodedURL).toString();
+            } catch (UnsupportedEncodingException e) {
+                return text;
+            }
+        }
+        return text;
+    }
+
+    private RawType downloadImage(String url) {
+        if (url != null && !url.isEmpty()) {
+            if (!IMAGE_CACHE.containsKey(url)) {
```

Comment:
I think you should use `computeIfAbsent` here:

```java
Map<String, String> test = new HashMap<>();
System.out.println(test.computeIfAbsent("test", k -> k + " -> new")); // prints test -> new
System.out.println(test.computeIfAbsent("test", k -> k + " -> ff")); // prints test -> new
System.out.println(test.computeIfAbsent("ffe", k -> k + " -> new")); // prints ffe -> new
```

Reply:
I would change it, but the `IMAGE_CACHE` is not a `HashMap`. It is an `ExpiringCacheMap` from ESH and that doesn't support `computeIfAbsent`, yet.

Reply:
That is a pity, because this is such a common pattern for such a cache.

Reply:
No, new question, why would you even bother checking, only to not override a cached value, then we maybe want to have something like `putIfAbsent` wdyt?

Reply:
Implementing the `Map` interface is not trivial as you might have found out when implementing it, we want something slightly different.

Reply:
I support your idea. Extending the `ExpiringCacheMap` to match the `Map` would be a very nice improvement. Further discussion in https://github.com/eclipse/smarthome/pull/4483.